packages/flutter/lib/src/services/asset_resolver.dart

```diff
@@ -0,0 +1,70 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:mojo/core.dart' as core;
+
+import 'asset_bundle.dart';
+
+/// Base class for asset resolvers.
+abstract class AssetResolver {
+  /// Return a resolved asset key for the asset named [name].
+  Future<String> resolve(String name);
+}
+
+/// Builder for a particular AssetResolver.
+///
+/// Allows AssetResolvers to be specified in widget build() functions.
+/// Subclasses should override operator== and hashCode so that equivalent
+/// configurations appear to be equal. Builder objects should also be
+/// cheap to construct.
+abstract class AssetResolverBuilder<T extends AssetResolver> {
+  const AssetResolverBuilder();
+  T buildResolver();
+}
```

Comment:
I don't understand why we need AssetResolverBuilder.  It seems like AssetResolvers should be cheap to create.  The ones you have in this patch just store parameters in their constructors.

Also, you don't need to re-use `operator==` to detect whether you need to dump the cache.  The pattern we use for other delegates of this type is to just have another function that returns a bool and that is given the "old" delegate as an argument.

The pattern is that the caller assumes it needs to flush its cache if the new AssetResolver has a different `runtimeType` from the old AssetResolver.  If the two have the same `runtimeType`, then it calls the function to see what it should do.

In your case, AssetResolver itself hold a cache, so you probably want to keep a reference to the current AssetResolver on the State object and replace it only when you dump your cache.